drivers/led_strip/tlc59731.c

```diff
@@ -0,0 +1,205 @@
+#define T_CYCLE_1 0x01 /* us */
+#define T_H0      (4 * T_CYCLE_0)
+#define T_H1      (8 * T_CYCLE_0)
+/* Threshould levels */
+#define HIGH      0x01
+#define LOW       0x00
+
+/* Write command */
+#define WR 0x3A
+
+struct tlc59731_cfg {
+	struct gpio_dt_spec sdi_gpio;
+	struct led_rgb led_rgb_data;
+};
+
+/*****************************/
+static inline int rgb_pulse(const struct gpio_dt_spec *led_dev)
+{
+	int32_t fret = 0;
+
+	fret = gpio_pin_set_dt(led_dev, HIGH);
+	if (fret != 0) {
+		return fret;
+	}
+
+	fret = gpio_pin_set_dt(led_dev, LOW);
+	if (fret != 0) {
+		return fret;
+	}
+
+	return fret;
+}
+
+static int rgb_eos(void)
+{
+	k_busy_wait(T_H0);
+	return 0;
+}
+
+static int rgb_gslatch(void)
+{
+	k_busy_wait(T_H1);
+
+	return 0;
+}
+
+static int rgb_write_bit(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	rgb_pulse(led_dev);
+
+	k_busy_wait(DELAY);
+
+	if (data) {
+		rgb_pulse(led_dev);
+		k_busy_wait(T_CYCLE_1);
+	} else {
+		k_busy_wait(T_CYCLE_0);
+	}
+
+	return 0;
+}
+
+static int rgb_write_data(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	int8_t idx = 7;
+
+	while (idx >= 0) {
+		rgb_write_bit(led_dev, data & BIT((idx--)));
+	}
+
+	return 0;
+}
+
+static int rgb_write_led(const struct gpio_dt_spec *led_dev, struct led_rgb *rgb_data)
+{
+	rgb_write_data(led_dev, WR);
+	rgb_write_data(led_dev, rgb_data->r);
+	rgb_write_data(led_dev, rgb_data->g);
+	rgb_write_data(led_dev, rgb_data->b);
+
+	rgb_eos();
+	rgb_gslatch();
+
+	return 0;
+}
+
+static int tlc59731_latch_and_write(const struct device *dev, struct led_rgb *pixel)
+{
+	int32_t err = 0;
+	const struct tlc59731_cfg *tlc_conf = dev->config;
+	const struct gpio_dt_spec *led_gpio = &tlc_conf->sdi_gpio;
+
+	err = rgb_write_led(led_gpio, pixel);
+	if (err) {
+		LOG_ERR("%s: GPIO write error", dev->name);
+	}
+
+	return err;
+}
+
+/**********************************/
+static int tlc59731_led_set_color(const struct device *dev, struct led_rgb *pixel)
+{
+	int32_t err = 0;
+
+	err = tlc59731_latch_and_write(dev, pixel);
+	if (err) {
+		LOG_ERR("%s: RGB write error: %d", dev->name, err);
+	}
+
+	return err;
+}
+
+static int tlc59731_gpio_update_rgb(const struct device *dev, struct led_rgb *pixels,
+				    size_t num_pixels)
+{
+	size_t i = 0;
+	int32_t err = 0;
+
+	for (i = 0; i < num_pixels; i++) {
+		err = tlc59731_led_set_color(dev, &pixels[i]);
+		if (err) {
+			break;
+		}
+	}
+
+	return err;
+}
+static int tlc59731_gpio_update_channels(const struct device *dev, uint8_t *channels,
+					 size_t num_channels)
+{
+	LOG_ERR("update_channels is not supported");
+	return -ENOTSUP;
+}
+
+static const struct led_strip_driver_api tlc59731_gpio_api = {
+	.update_rgb = tlc59731_gpio_update_rgb,
+	.update_channels = tlc59731_gpio_update_channels,
+};
+
+#define TLC59731_DEVICE(i)                                                                         \
+	static int tlc59731_gpio_##i##_init(const struct device *dev)                              \
```

Comment:
I don't think this function needs to be defined inside the macro, there is no difference between its different instances.